### ISSUE TITLE
How to dynamically load different xml settings files

### ISSUE THREAD

**Issue by olivercrush** [kr15h/ofxPiMapper#145]
Hi,

I would like to know how I can dynamically change the xml settings file. I would basically like to change some surfaces and sources when I press a button. Is there a way to easily do that ? I saw that there is a preset system too but I'm not really sure what it does and how it works.

Thank you in advance for your response, you did a great job on this project !

**Comment by kr15h** [comment]
You mean dynamically load another xml file? There is the function `ofxPiMapper::loadProject(std::string filename)`. Try that. There might be issues, since the `ofxOMXPlayer` in the background might not unload video files properly and it would eat up memory. If you use it with images, should work fine. Got to do something with this OMX player...

**Comment by olivercrush** [comment]
Yes I tried calling loadProject in my ofApp.cpp, but it freezes everything and I have to reboot my pi. I can't even get back to the command interface to see the debug messages so it's hard to tell where exactly the problem occurs. And I don't even use videos in my mapping, there are just images

**Comment by kr15h** [comment]
Yes, I guess that is a valid issue. The thing with pressets is that you can store multiple projects in one xml file. You could try that approach. You just have to add another `<surfaces>...</surfaces>` block after the one that is there. That is for now.

**Comment by olivercrush** [comment]
Oh ok thanks, I will look into it. I guess I can switch between presets with the setPreset() method ?

**Comment by kr15h** [comment]
Yes.

**Comment by pietrondo** [comment]
is possible to have this "function" in the example? (so if i press 7 i load preset 1, 8 load preset 2 ecc..?)
thx

**Comment by olivercrush** [comment]
Here, I can give you my ofApp's keyPressed method for an example. You just have to have several <surfaces></surfaces> with your configurations in your ofxpimapper.xml

```
void ofApp::keyPressed(int key) {

	switch (key) {
		case 'q':
			mapper.setPreset(0);
			break;

		case 'w':
			mapper.setPreset(1);
			break;

		case 'e':
			mapper.setPreset(2);
			break;

		case 'r':
			mapper.setPreset(3);
			break;

		case 't':
			mapper.setPreset(4);
			break;

		case 'y':
			mapper.setPreset(5);
			break;

		case 'u':
			mapper.setPreset(6);
			break;

		case 'i':
			mapper.setPreset(7);
			break;

		case 'o':
			mapper.setPreset(8);
			break;

		case 'p':
			mapper.setPreset(9);
			break;

		case 'a':
			mapper.setPreset(10);
			break;

		case 's':
			mapper.setPreset(11);
			break;

		default:
			mapper.keyPressed(key);
			break;
	}
	usleep(10);
}
```

**Comment by kr15h** [comment]
Yes, this is how it is meant to be. One can override internal ofxPiMapper shortcuts.

**Comment by pietrondo** [comment]
yes... a feature could be to have a "config file" in witch are defined the
keys and could be reconfigured.
thx


Pietro Capriata

http://www.pietrocapriata.me


CONFIDENTIALITY DISCLAIMER
Le informazioni contenute in questo messaggio di posta elettronica e negli
allegati sono riservate e confidenziali e ne sono vietate la lettura,
l'uso, la copia, la comunicazione e la diffusione in qualunque modo
eseguite. Qualora lei non fosse la persona destinataria del messaggio, la
invitiamo ad eliminarlo, dando gentilmente comunicazione al mittente
tramite e-mail di ritorno.

The information contained in this email message and any files transmitted
with it are confidential and privileged and any reading, processing,
distribution or copy of this material is strictly prohibited, in any form.
If you are not the intended recipient of this message, please immediately
delete it, giving the relevant communication to the sender by reply e-mail.

**Comment by kr15h** [comment]
You can also take a look at the [example_pocketvj ofApp.cpp](https://github.com/kr15h/ofxPiMapper/blob/master/example_pocketvj/src/ofApp.cpp) and use that to remap your keys. But I would prefer the config file solution proposed by @pietrondo myself.
